### PR TITLE
Rewind the network state to the epoch with the best validation set.

### PR DESCRIPTION
**Pull request by mrkrause**
I'm not sure if abusing __get__state like this is smart or a terrible idea.

### CODE REVIEW

File: src/net.py [hunk 1]
```diff
@@ -40,8 +40,9 @@ class Network(object):
 
     def __init__(self, name, dimensions, input_var, y, config,
                  input_network=None, num_classes=None, activation='rectify',
-                 pred_activation='softmax', optimizer='adam',
+                 pred_activation='softmax', optimizer='adam', stopping_rule='best_validation',
```
Comment (rfratila) [inline src/net.py]: I think we should default it to None so you have to explicitly specify if you want to use early stopping

File: src/net.py [hunk 2]
```diff
@@ -59,7 +60,9 @@ def __init__(self, name, dimensions, input_var, y, config,
             activation: layer activation function
             pred_activation: the classifying layer activation
             optimizer: which optimizer to usem as the learning function
-            learning_rate: the initial learning rate
+            learning_rate: the initial learning rate,
+            stopping_rule: (stub) If best_validation, use best e[
```
Comment (rfratila) [inline src/net.py]: is the e[ a typo?

File: src/net.py [hunk 3]
```diff
@@ -588,10 +606,16 @@ def train(self, epochs, train_x, train_y, val_x, val_y,
                         )
                         # print ('to {}'.format(self.learning_rate))
                     self.minibatch_iteration += 1
+
                 train_error, train_accuracy = self.validator(train_x, train_y)
                 validation_error, validation_accuracy = self.validator(val_x,
                                                                        val_y)
 
+                if self.stopping_rule == 'best_validation' and validation_error < best_error:
+                    best_state = self.__getstate__()
```
Comment (rfratila) [inline src/net.py]: I think to save more memory, it might be a better idea to just save the parameters using a similar command to ```net_parameters = np.array(lasagne.layers.get_all_param_values(self.layers,**{self.name: True}))``` found in the __getstate__ function. using just __getstate__ will copy over other parts of the class that are not necessary.

File: src/net.py [hunk 4]
```diff
@@ -622,11 +646,18 @@ def train(self, epochs, train_x, train_y, val_x, val_y,
...
+
+
         except KeyboardInterrupt:
             print ("\n\n**********Training stopped prematurely.**********\n\n")
         finally:
             self.timestamp = get_timestamp()
 
+            if self.stopping_rule == 'best_validation':
+                print("STOPPING RULE: Rewinding to epoch {} which had the lowest validation error: {}".format(best_epoch, best_error))
+                self.__setstate__(best_state)
```
Comment (rfratila) [inline src/net.py]: If we were to just save the weights of the network (using my previous recommendation) then we can load them back into the network here with a command similar to ```lasagne.layers.set_all_param_values(self.layers, weights, **{self.name: True})```. I think this would run faster. thoughts?

File: src/net.py [hunk 5]
```diff
@@ -125,12 +129,14 @@ def __init__(self, name, dimensions, input_var, y, config,
...
             lasagne.layers.get_output(self.network, deterministic=True))
         self.record = None
+
         try:
             self.timestamp
         except AttributeError:
             self.timestamp = get_timestamp()
         self.minibatch_iteration = 0
 
+
```
Comment (rfratila) [inline src/net.py]: Just a general comment but I don't think there need to be two spaces between functions that belong to a class. I think that applies to only functions outside of a class.

File: src/net.py [hunk 6]
```diff
@@ -588,10 +606,16 @@ def train(self, epochs, train_x, train_y, val_x, val_y,
                         )
                         # print ('to {}'.format(self.learning_rate))
                     self.minibatch_iteration += 1
+
                 train_error, train_accuracy = self.validator(train_x, train_y)
                 validation_error, validation_accuracy = self.validator(val_x,
                                                                        val_y)
 
+                if self.stopping_rule == 'best_validation' and validation_error < best_error:
```
Comment (rfratila) [inline src/net.py]: I think there should be a heuristic for determining early stopping by looking at the general trend of the validation error and see if it starts to further itself from the training error.

File: src/net.py (hunk 1 above)
Comment (mrkrause) [reply src/net.py]: Hmmm....I think 'best on the validation set' should be the default since it will typically have less generalization error than the probably-overfit 'best on the training set'. 

I agree that if/when we add other early stopping options, they should require opting in though, since those could presumably cost you performance.

File: src/net.py (hunk 6 above)
Comment (mrkrause) [reply src/net.py]: I've been looking for something relatively principled, but yeah.... People seem to run for a few iterations after hitting a local minima, but the value of "few" seems to be pulled out of thin air. Can add something in.

File: src/net.py (hunk 3 above)
Comment (mrkrause) [reply src/net.py]: I figured this was slightly safer than copying the weights and performance data over (since we might add stuff and forget to add it to the copier), but it's your call.

File: src/net.py (hunk 1 above)
Comment (rfratila) [reply src/net.py]: Fair enough, we can keep it that way. not sure how this affects saving and loading functionality. Would you be able to run some tests with that in mind?

File: src/net.py (hunk 6 above)
Comment (rfratila) [reply src/net.py]: if you include some variable, epsilon, that acts as a sensitivity threshold for how much the validation needs to deviate from the training in order to stop that would make it more general purpose. thoughts?

File: src/net.py (hunk 3 above)
Comment (rfratila) [reply src/net.py]: Im just worried about the additional memory overhead that this would introduce after training for a while.
